common/ferrostar/src/navigation_controller/mod.rs

```diff
@@ -37,16 +37,114 @@ pub struct NavigationController {
     config: NavigationControllerConfig,
 }
 
+/// Core interface for navigation functionalities.
+///
+/// This trait defines the essential operations for a navigation state manager.
+/// This lets us build additional layers (e.g. event logging)
+/// around [`NavigationController`] a composable manner.
+#[cfg_attr(feature = "uniffi", uniffi::export)]
+pub trait Navigator: Send + Sync {
+    fn get_initial_state(&self, location: UserLocation) -> TripState;
+    fn advance_to_next_step(&self, state: &TripState) -> TripState;
+    fn update_user_location(&self, location: UserLocation, state: &TripState) -> TripState;
+}
+
+/// Creates a new navigation controller for the given route and configuration.
+///
+/// It returns an Arc-wrapped trait object implementing `Navigator`.
+/// If `should_record` is true, it creates a controller with event recording enabled.
+#[cfg_attr(feature = "uniffi", uniffi::export)]
+pub fn create_navigator(
+    route: Route,
+    config: NavigationControllerConfig,
+    should_record: bool,
+) -> Arc<dyn Navigator> {
+    if should_record {
+        // Creates a navigation controller with a wrapper that records events.
+        // TODO: Currently just returns the regular controller
+        Arc::new(NavigationController { route, config })
+    } else {
+        // Creates a normal navigation controller.
+        Arc::new(NavigationController { route, config })
+    }
+}
+
 #[cfg_attr(feature = "uniffi", uniffi::export)]
 impl NavigationController {
     #[cfg_attr(feature = "uniffi", uniffi::constructor)]
     /// Create a navigation controller for a route and configuration.
     pub fn new(route: Route, config: NavigationControllerConfig) -> Self {
         Self { route, config }
     }
+}
+
+/// Shared functionality for the navigation controller that is not exported by uniFFI.
```

Comment:
Not your comment, but since it showed up in the diff... ;)

```suggestion
/// Shared functionality for the navigation controller that is not exported by UniFFI.
```